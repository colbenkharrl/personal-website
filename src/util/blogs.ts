import glob from 'fast-glob';
import { StaticImageData } from 'next/image';

interface Blog {
  title: string;
  description: string;
  author: string;
  date: string;
}

export interface BlogWithSlug extends Blog {
  slug: string;
  opengraphImage: StaticImageData;
}

async function importBlog(blogFilename: string): Promise<BlogWithSlug> {
  const { blog } = (await import(`../app/blog/${blogFilename}`)) as {
    default: React.ComponentType;
    blog: Blog;
  };

  const slug = blogFilename.replace(/(\/page)?\.mdx$/, '');

  const { default: opengraphImage } = (await import(
    `../app/blog/${slug}/opengraph-image.png`
  )) as {
    default: StaticImageData;
  };

  return {
    slug,
    opengraphImage,
    ...blog,
  };
}

export async function getAllBlogs() {
  const blogFilenames = await glob('*/page.mdx', {
    cwd: './src/app/blog',
  });

  const blogs = await Promise.all(blogFilenames.map(importBlog));

  return blogs.sort((a, z) => +new Date(z.date) - +new Date(a.date));
}
