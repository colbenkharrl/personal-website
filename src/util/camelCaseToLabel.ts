/**
 * Converts a camelCase string to a formatted label with spaces and each word capitalized.
 * @param camelCaseStr The camelCase string to be transformed.
 * @returns The formatted label.
 */
export const camelCaseToLabel = (camelCaseStr: string): string => {
  // Adjust regex to separate numbers and uppercase letters, treat sequences of numbers followed by lowercase as separate words
  const spacedStr = camelCaseStr
    .replace(/([0-9]+|[A-Z])/g, ' $1')
    .replace(/(\d+)([a-z])/g, '$1 $2');
  // Trim any leading space and capitalize each word
  return spacedStr.trim().replace(/\b\w/g, (char) => char.toUpperCase());
};
