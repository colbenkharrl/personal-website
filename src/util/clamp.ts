export function clamp(number: number, a: number, b: number) {
  let min = Math.min(a, b);
  let max = Math.max(a, b);
  return Math.min(Math.max(number, min), max);
}
