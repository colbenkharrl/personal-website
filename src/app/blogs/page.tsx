import { type Metadata } from 'next';

import { SimpleLayout } from '@/components/SimpleLayout';
import { getAllBlogs } from '@/util/blogs';
import { Blog } from '@/components/Blog';

export const metadata: Metadata = {
  title: 'Blog',
  description: 'My scattered thoughts about random things.',
};

export default async function BlogsIndex() {
  const blogs = await getAllBlogs();

  return (
    <SimpleLayout
      title="My scattered thoughts about random things."
      intro="I sometimes write down my ideas or document my experiences as I Roomba my way through my professional and personal life. A subset of those writings end up being published here for likely no one to enjoy. So, enjoy!"
    >
      <div className="md:border-l md:border-zinc-100 md:pl-6 md:dark:border-zinc-700/40">
        <div className="flex max-w-3xl flex-col space-y-16">
          {blogs.map((blog) => (
            <Blog key={blog.slug} blog={blog} />
          ))}
        </div>
      </div>
    </SimpleLayout>
  );
}
