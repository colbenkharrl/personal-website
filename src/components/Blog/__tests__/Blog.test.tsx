import { render, screen } from '@testing-library/react';
import { Blog } from '../';
import { BlogWithSlug } from '@/util/blogs';

it('Blog matches snapshot (regression test)', () => {
  const testBlog: BlogWithSlug = {
    slug: 'test-blog',
    title: 'Test Blog',
    description: 'Test blog description.',
    author: 'Matt Kharrl',
    date: '2023-11-09',
  };

  render(<Blog blog={testBlog} />);
  expect(screen.getByTestId('Card')).toMatchSnapshot(
    'Full component snapshot.',
  );
});
