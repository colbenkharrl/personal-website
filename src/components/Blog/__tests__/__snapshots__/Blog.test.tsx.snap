// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blog matches snapshot (regression test): Full component snapshot. 1`] = `
<article
  class="group relative flex items-start flex flex-col sm:flex-row flex-row"
  data-testid="Card"
>
  <img
    alt=""
    class="mb-6 aspect-video w-full shrink-0 grow-0 rounded-lg object-cover sm:mb-0 sm:mr-6 sm:w-72"
    data-nimg="1"
    decoding="async"
    height="40"
    loading="lazy"
    src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
    srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
    style="color: transparent;"
    width="40"
  />
  <div
    class="flex flex-col"
  >
    <h2
      class="text-base font-semibold tracking-tight text-zinc-800 dark:text-zinc-100"
    >
      <div
        class="absolute -inset-x-4 -inset-y-6 z-0 scale-95 bg-zinc-50 opacity-0 transition group-hover:scale-100 group-hover:opacity-100 dark:bg-zinc-800/50 sm:-inset-x-6 sm:rounded-2xl"
      />
      <a
        href="/blog/test-blog"
      >
        <span
          class="absolute -inset-x-4 -inset-y-6 z-20 sm:-inset-x-6 sm:rounded-2xl"
        />
        <span
          class="relative z-10"
        >
          Test Blog
        </span>
      </a>
    </h2>
    <time
      class="relative z-10 order-first mb-3 flex items-center text-sm text-zinc-400 dark:text-zinc-500 pl-3.5"
      datetime="2023-11-09"
    >
      <span
        aria-hidden="true"
        class="absolute inset-y-0 left-0 flex items-center"
      >
        <span
          class="h-4 w-0.5 rounded-full bg-zinc-200 dark:bg-zinc-500"
        />
      </span>
      November 9, 2023
    </time>
    <p
      class="relative z-10 mt-2 text-sm text-zinc-600 dark:text-zinc-400"
    >
      Test blog description.
    </p>
    <div
      aria-hidden="true"
      class="relative z-10 mt-4 flex items-center text-sm font-medium text-orange-500"
    >
      Read blog post
      <svg
        aria-hidden="true"
        class="ml-1 h-4 w-4 stroke-current"
        fill="none"
        viewBox="0 0 16 16"
      >
        <path
          d="M6.75 5.75 9.25 8l-2.5 2.25"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="1.5"
        />
      </svg>
    </div>
  </div>
</article>
`;
